### ISSUE TITLE
[Feature] MVVM Toolkit vNext: source generators! 🚀

### ISSUE THREAD

**Issue by Sergio0694** [CommunityToolkit/dotnet#8]
The `Microsoft.Toolkit.Mvvm` package (aka MVVM Toolkit) is finally out, so it's time to start planning for the next update 😄
This issue is meant to be for tracking planned features, discuss them and possibly propose new ideas well (just like in https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3428).

One aspect I want to investigate for this next release, which I think makes sense for the package, is support for [Roslyn source generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/). Support for these would be added through a secondary project, ie. `Microsoft.Toolkit.Mvvm.SourceGenerators`, that would be shipped **together with the MVVM Toolkit**, but then added to consuming projects as a development dependency, ie. as an analyzer. They would be included in the same NuGet package, so it wouldn't be possible to just reference the source generator directly (which wouldn't make sense anyway). This is the same setup I'm using in my own library [ComputeSharp](https://github.com/Sergio0694/ComputeSharp), in fact I'm applying a lot of the things I learnt while working on that project to the MVVM Toolkit as well 🥳

There are ~~two~~ three main aspects that can be investigated through support for source generators:

- Better extensibility and composition (opt-in)
- Less verbosity (opt-in)
- Better performance (always on)

Here's some more details about what I mean by these two categories exactly.

## Better extensibility

There's one "problem" with C# that has been brought up by devs before while using the MVVM Toolkit and in general (eg. [here](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3804)): lack for multiple inheritance. While that makes perfect sense and there's plenty of valid reasons why that's the case, the fact remains that it makes things sometimes inconvenient, when eg. you want/need to inherit from a specific type but then still want to use the features exposed by other classes in the MVVM Toolkit. The solution? Source generators! 🚀

For now I'm thinking about adding the following attributes to the package:
- `[INotifyPropertyChanged]`
- `[ObservableObject]`
- `[ObservableRecipient]`

The way they work is that they let you annotate a type and then rely on the generator injecting all the APIs from those types automatically, so you don't need to worry about it and it's like you were effectively having multiple inheritance. Here's an example:

```csharp
[ObservableObject]
partial class MyViewModel : SomeOtherClass
{
}
```

This class now inherits from `SomeOtherClass`, **but** it still has all the same APIs from `ObservableObject`! This includes the `PropertyChanged` and `PropertyChanging` events, the methods to raise them and all the additional helper methods!

`[ObservableRecipient]` does the same but copying members from `ObservableRecipient` (eg. this could be used to effectively have a type that is both a validator but also a recipient), and `[INotifyPropertyChanged]` instead offers minimal support just for `INotifyPropertyChanged`, with optionally also the ability to include additional helpers or not. This approach and the different attributes offer maximum flexibility for users to choose the best way to construct their architecture without having to compromise between what APIs to use from the MVVM Toolkit and how they want/have to organize their type hierarchy. 🎉

> **NOTE:** this category is marked as "opt-in" because the attributes are completely optional. Not using them will have no changes at all on the behavior of the toolkit, so developers just wanting to inherit from the base types in the library as usual will absolutely still be able to do so. This just gives consumers more flexibility depending on their exact use case scenario.

## Less verbosity

This was first suggested by @michael-hawker in [this comment](https://github.com/windows-toolkit/WindowsCommunityToolkit/issues/3872#issuecomment-804251339), the idea is to also provide helpers to reduce the code verbosity in simple cases, such as when defining classic observable properties. For now I've added these attributes:

- `[ObservableProperty]`
- `[AlsoNotifyFor]`
- `[ICommand]`

The first two can be used to easily declare observable properties, by annotating a field. `[ObservableProperty]` will create the code necessary to implement the property itself, whereas `[AlsoNotifyFor]` will customize the generated code by adding extra notification events (ie. calls to `OnPropertyChanged`) for properties whose value depends on the property being updated.

Here's an example of how these two attributes can be used together:

<details>
    <summary><b>Viewmodel definition:</b></summary>
<br/>

```csharp
public sealed partial class PersonViewModel : ObservableObject
{
    [ObservableProperty]
    [AlsoNotifyFor(nameof(FullName))]
    private string name;

    [ObservableProperty]
    [AlsoNotifyFor(nameof(FullName))]
    private string surname;

    public string FullName => $"{Name} {Surname}";
}
```
</details>

<details>
    <summary><b>Generated code:</b></summary>
<br/>

```csharp
public sealed partial class PersonViewModel
{
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public string Name
    {
        get => name;
        set
        {
            if (!global::System.Collections.Generic.EqualityComparer<string>.Default.Equals(name, value))
            {
                OnPropertyChanging();
                name = value;
                OnPropertyChanged();
                OnPropertyChanged("FullName");
            }
        }
    }

    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public string Surname
    {
        get => surname;
        set
        {
            if (!global::System.Collections.Generic.EqualityComparer<string>.Default.Equals(surname, value))
            {
                OnPropertyChanging();
                surname = value;
                OnPropertyChanged();
                OnPropertyChanged("FullName");
            }
        }
    }
}
```
</details>

There is also a brand new `[ICommand]` attribute type, which can be used to easily create command properties over methods, by leveraging the relay command types in the MVVM Toolkit. The way this works is simple: just write a method with a valid signature for either `RelayCommand`, `RelayCommand<T>`, `AsyncRelayCommand` or `AsyncRelayCommand<T>` and add the `[ICommand]` attribute over it - the generator will create a lazily initialized property with the right command type that will automatically wrap that method. Cancellation tokens for asynchronous commands are supported too! 🚀

Here's an example of how this attribute can be used with four different command types:

<details>
    <summary><b>Viewmodel definition:</b></summary>
<br/>

```csharp
public sealed partial class MyViewModel
{
    [ICommand]
    private void Greet()
    {
    }

    [ICommand]
    private void GreetUser(User user)
    {
    }

    [ICommand]
    private Task SaveFileAsync()
    {
        return Task.CompletedTask;
    }

    [ICommand]
    private Task LogUserAsync(User user)
    {
        return Task.CompletedTask;
    }
}
```
</details>

<details>
    <summary><b>Generated code:</b></summary>
<br/>

```csharp
public sealed partial class MyViewModel
{
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    private global::Microsoft.Toolkit.Mvvm.Input.RelayCommand? greetCommand;

    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public global::Microsoft.Toolkit.Mvvm.Input.IRelayCommand GreetCommand => greetCommand ??= new global::Microsoft.Toolkit.Mvvm.Input.RelayCommand(new global::System.Action(Greet));

    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    private global::Microsoft.Toolkit.Mvvm.Input.RelayCommand<global::MyApp.User>? greetUserCommand;

    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public global::Microsoft.Toolkit.Mvvm.Input.IRelayCommand<global::MyApp.User> GreetUserCommand => greetUserCommand ??= new global::Microsoft.Toolkit.Mvvm.Input.RelayCommand<global::UnitTests.Mvvm.Test_ICommandAttribute.User>(new global::System.Action<global::UnitTests.Mvvm.Test_ICommandAttribute.User>(GreetUser));
    
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    private global::Microsoft.Toolkit.Mvvm.Input.AsyncRelayCommand? saveFileAsyncCommand;

    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public global::Microsoft.Toolkit.Mvvm.Input.IAsyncRelayCommand SaveFileAsyncCommand => saveFileAsyncCommand ??= new global::Microsoft.Toolkit.Mvvm.Input.AsyncRelayCommand(new global::System.Func<global::System.Threading.Tasks.Task>(SaveFileAsync));
    
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    private global::Microsoft.Toolkit.Mvvm.Input.AsyncRelayCommand<global::MyApp.User>? logUserAsyncCommand;
    
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ICommandGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public global::Microsoft.Toolkit.Mvvm.Input.IAsyncRelayCommand<global::MyApp.User> LogUserAsyncCommand => logUserAsyncCommand ??= new global::Microsoft.Toolkit.Mvvm.Input.AsyncRelayCommand<global::UnitTests.Mvvm.Test_ICommandAttribute.User>(new global::System.Func<global::UnitTests.Mvvm.Test_ICommandAttribute.User, global::System.Threading.Tasks.Task>(LogUserAsync));
}
```
</details>

## Better performance

Another area that I want to investigate with source generators is possibly getting some performance improvemeents by removing reflection where possible. Now, the MVVM Toolkit is already quite light on reflection (as it was designed with that in mind, especially the messenger types), but I think there might be a few places where things could still be improved with source generators. For instance, [this method](https://github.com/windows-toolkit/WindowsCommunityToolkit/blob/2c04c96cdd9e9c574f4a7a5c33d6b3a7ae8546ac/Microsoft.Toolkit.Mvvm/Messaging/IMessengerExtensions.cs#L93-L94) uses quite a bit of reflection.

We could keep this for compatibility and also as a "fallback" implementation, but then we could have the source generator emit a type-specific version of this method with all the necessary handlers already specified, with no reflection. We'd just need to generate the appropriate method in the consuming assembly, and then the C# compiler would automatically pick that one up due to how overload resolution works (since the `object recipient` in the original method is less specific than a `MyViewModel recipient` parameter that the generated method would have). Still haven't done a working proof of concept for this point specifically, but it's next on my list and will update as soon as that's done too, just wanted to open this issue in the meantime to start gathering feedbacks and discuss ideas 🙂

**EDIT:** I've now added a generator that will create a method for this for all types implementing `IRecipient<T>`:

```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
#pragma warning disable
namespace Microsoft.Toolkit.Mvvm.Messaging.__Internals
{
    internal static partial class __IMessengerExtensions
    {
        [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
        [global::System.Obsolete("This method is not intended to be called directly by user code")]
        public static global::System.Action<IMessenger, object> CreateAllMessagesRegistrator(global::MyApp.MyViewModel _)
        {
            static void RegisterAll(IMessenger messenger, object obj)
            {
                var recipient = (global::MyApp.MyViewModel)obj;
                messenger.Register<global::MyApp.MessageA>(recipient);
                messenger.Register<global::MyApp.MessageB>(recipient);
            }

            return RegisterAll;
        }

        [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
        [global::System.Obsolete("This method is not intended to be called directly by user code")]
        public static global::System.Action<IMessenger, object, TToken> CreateAllMessagesRegistratorWithToken<TToken>(global::MyApp.MyViewModel _)
            where TToken : global::System.IEquatable<TToken>
        {
            static void RegisterAll(IMessenger messenger, object obj, TToken token)
            {
                var recipient = (global::MyApp.MyViewModel)obj;
                messenger.Register<global::MyApp.MessageA, TToken>(recipient, token);
                messenger.Register<global::MyApp.MessageB, TToken>(recipient, token);
            }

            return RegisterAll;
        }
    }
}
```

This is then now picked up automatically when `RegisterAll` is called, so that the LINQ expression can be skipped entirely.
There are two generated methods so that the non-generic one can be used in the more common scenario where a registration token is not used, and that completely avoids runtime-code generation of all sorts and also more reflection (no more `MakeDynamicMethod`), making it particularly AOT-friendly 😄

**EDIT 2:** I've applied the same concept to the other place where I was using compiled LINQ expressions too, that is the `ObservableValidator.ValidateAllProperties` method. We now have a new generator that will process all types inheriting from `ObservableValidator`, and create helper methods like this that will then be loaded at runtime by the MVVM Toolkit as above:

```csharp
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.
#pragma warning disable
namespace Microsoft.Toolkit.Mvvm.ComponentModel.__Internals
{
    [global::System.CodeDom.Compiler.GeneratedCode("Microsoft.Toolkit.Mvvm.SourceGenerators.ObservableValidatorValidateAllPropertiesGenerator", "7.0.0.0")]
    [global::System.Diagnostics.DebuggerNonUserCode]
    [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
    [global::System.Obsolete("This type is not intended to be used directly by user code")]
    internal static partial class __ObservableValidatorExtensions
    {
        [global::System.ComponentModel.EditorBrowsable(global::System.ComponentModel.EditorBrowsableState.Never)]
        [global::System.Obsolete("This method is not intended to be called directly by user code")]
        public static global::System.Action<object> CreateAllPropertiesValidator(global::MyApp.PersonViewModel _)
        {
            static void ValidateAllProperties(object obj)
            {
                var instannce = (global::MyApp.PersonViewModel)obj;
                __ObservableValidatorHelper.ValidateProperty(instance, instance.Name, nameof(instance.Name));
                __ObservableValidatorHelper.ValidateProperty(instance, instance.Age, nameof(instance.Age));
            }

            return ValidateAllProperties;
        }
    }
}
```

When the source generators are used, the MVVM Toolkit is now 100% without dynamic code generation! 🎉

## Tracking changes so far

- [X] `[INotifyPropertyChangedAttribute]`
- [X] `[ObservableObjectAttribute]`
- [X] `[ObservableValidatorAttribute]` 
- [X] Additional support for generated `RegisterAll` method when possible
- [X] Additional support for generated `ValidateAllProperties` method when possible
- [X] `[ObservableProperty]`
- [X] `[AlsoNotifyChangeFor]`
- [X] `[ICommand]`
- [x] `CanExecute` property for `[ICommand]`
- [x] `[AlsoNotifyCanExecuteFor]`
- [x] Switch to incremental generators

Feedbacks and feature ideas are welcome! 🙌

**Comment by kmgallahan** [comment]
@Sergio0694 You might have missed [my comment above](https://github.com/CommunityToolkit/dotnet/issues/8#issuecomment-965255311) while transferring this issue. If not, apologies, no rush for reply.

**Comment by Sergio0694** [comment]
Hey @kmgallahan, sorry for the delay 😄
Glad to hear you like this feature so far, that's awesome!

> "Really liking `[ICommand]` and looking forward to `[CanExecute]`"

Was planning to do some work on the MVVM Toolkit today, so might go implement just that 😋

Regarding your questions about the differences with Fody, there's some pretty fundamental differences between that and the MVVM Toolkit (which uses source generators instead of IL weaving), which are by design. Going through your questions:

1. This is not feasible, unfortunately. The reason for this is that source generators are (by design) addictive-only, so they cannot modify in any way existing code, but only add new one. Because of this, you can't just set a flag to have every autoproperty automatically become observable like you do with Fody, as that'd require IL changes to the property accessors.
2. This is also not doable, for the same reason. The closest you can get is to use `[ObservableProperty]` on a field.
3. This would be significantly more complex (I imagine you're thinking about doing something like crawling the syntax tree for a property getter and identify other referenced properties?), and might not necessarily result in the exact behaviors developers might want in all scenarios. Using an attribute is more explicit, simpler to implement, and gives you more control.
4. It shouldn't cause issues with Fody in theory, other that maybe in hot reload scenarios. For normal builds, the source generators would get invoked first, it would generate all additional code required, then Fody would go rewrite the final compiled assembly, which would also include all code produced by the generators. As in, from Fody's point of view, the code produced by these generators should be just like any other code normally generated by the C# compiler.

Hope this helps! :)

**Comment by Sergio0694** [comment]
Hey everyone! 👋
We just published a new preview on NuGet, it's version 8.0.0 Preview 1.
It includes all the recent changes to the MVVM Toolkit I've been doing, such as:

- Switching to incremental source generators
- All new attributes mentioned in this post
- Ability to set `CanExecute` for commands
- Ability to configure concurrent execution for async commands automatically

Plus a lot of fixes (eg. they should work with generic viewmodels too now, etc.).
If you try it out, please let us know how it goes and if you find any issues! 😄

**Comment by virzak** [comment]
Works great for synchronous commands except one tiny issue, that is probably unrelated to this project.
Go To Definition (F12) in the xaml file results in `Cannot navigate to definition.` When ICommand property is written manually, navigation works no problem.

Asynchronous commands are held by CommunityToolkit/WindowsCommunityToolkit#3961

So for my project roughly half of the commands can be converted. Still worth it despite the inconsistency in approaches, IMO. But resolving that issue would be great.

**Comment by Sergio0694** [comment]
> "Go To Definition (F12) in the xaml file results in Cannot navigate to definition. When ICommand property is written manually, navigation works no problem."

I suspect that might just be an issue with IntelliSense in VS. Things should get better with VS2022 17.2 (eg. see https://github.com/dotnet/roslyn/issues/51810). I have the same problem with my generators in ComputeSharp, where Visual Studio will often just display red squiggles all over the place despite code building and running just fine. We probably just need to wait for IDE support to improve on this, I don't think it's something we can fix on our end with changes to the generator.

> "Asynchronous commands are held by CommunityToolkit/WindowsCommunityToolkit#3961"

Yup, looking into that is definitely on my roadmap 🙂

**Comment by wokhan** [comment]
> Plus a lot of fixes (eg. **they should work with generic viewmodels too now**, etc.). If you try it out, please let us know how it goes and if you find any issues! 😄

Thank you @Sergio0694 ! (and other contributors, of course). I was waiting for this one to clean up some classes. I'll test the other new features asap!

**Comment by proudust** [comment]
I noticed that `ICommand.CanExecute` in v8.0.0 Preview 1 does not support the following patterns.
Is it possible to support this?

```cs
[ObservableProperty]
private bool _canInvoke;

[ICommand(CanExecute = nameof(CanInvoke))]
private void Invoke()
//           ^^^^^^
// MVVMTK0009: The CanExecute name must refer to a valid member, but "CanInvoke" has no matches in type Class
{

}
```

**Comment by Sergio0694** [comment]
Mmmh that _should_ work, yes, but it's very possible I forgot to implement that bit ahah. I'll take a look this week, thanks! 😄

**Comment by wokhan** [comment]
> I noticed that `ICommand.CanExecute` in v8.0.0 Preview 1 does not support the following patterns.
> Is it possible to support this?
> 
> ```cs
> [ObservableProperty]
> private bool _canInvoke;
> 
> [ICommand(CanExecute = nameof(CanInvoke))]
> private void Invoke()
> //           ^^^^^^
> // MVVMTK0009: The CanExecute name must refer to a valid member, but "CanInvoke" has no matches in type Class
> {
> 
> }
> ```

Hi, I guess you did that already, but Class is partial, right? And did you try a rebuild _before_ using nameof with the generated property? Sometimes it fails finding them on my side, but rebuilding first fixes this...

**Comment by Sergio0694** [comment]
@proudust I've fixed this, it's now tracked in https://github.com/CommunityToolkit/dotnet/pull/106. Feel free to try out a nightly build from the PR feed if you want to double check that it works for you, or to unblock you until that's generally available. Thank you for reporting this! 😄

**Comment by proudust** [comment]
@Sergio0694 Using the NuGet package created by checking out main (0936b58854f1d2a1120fff2ac82d21eb430db95e), we have verified that the above pattern is now supported. Thanks for the quick response.

**Comment by timunie** [comment]
Hi @Sergio0694 , 

is it possible to add xml-comments to generated properties? I just like it if I hover over any property and can read the docs about them. It's not a show stopper, but might be a good addition. 

Thank you and happy coding 
Tim

**Comment by Sergio0694** [comment]
Posting here to notify folks in this conversation as well.
There's a couple proposals in other issues you might be interested in 😄

Partial methods for observable properties: https://github.com/CommunityToolkit/dotnet/issues/117
Cancel commands: https://github.com/CommunityToolkit/dotnet/issues/118

The first would essentially result in `OnPropertyChanging/Changed` partial methods being generated when using `[ObservableProperty]`, to give devs more flexibility. The second would allow the commands generator to also easily create a command to cancel another async command. Feedbacks welcome! 😊

**Comment by Sergio0694** [comment]
Hey folks!
I published 8.0.0-Preview2 yesterday as many of you might've seen, let me know what you think!
The changelog is in the release notes in this repo 😄

Small disclaimer: I accidentally broke notifications for `CanExecute` for async relay commands, sorry about that 🤣

I'll get started with work for the next preview, keep the feedback coming! It's being incredibly useful! 😊

**Comment by timunie** [comment]
> I accidentally broke notifications for CanExecute for async relay commands, sorry about that 🤣

Oh dear, this one is hitting me now. Should I use an earlier preview or is there any nightly build where is is fixed? 

Anyway, great work as always. It's really fun to use :-)

**Comment by Sergio0694** [comment]
You can add our feed for getting builds from main (`https://pkgs.dev.azure.com/dotnet/CommunityToolkit/_packaging/CommunityToolkit-MainLatest/nuget/v3/index.json`) and then just grab whatever the latest version is from there (it'll be `8.0.0-build.<NUMBER>`). That'll fix the issue 🙂

**Comment by timunie** [comment]
Yes it did. thank you so much for all your work and the very very fast reply :-)

**Comment by CyberBotX** [comment]
I was migrating to use this toolkit's `RelayCommand` and `AsyncRelayCommand` to replace my usage of PostSharp's XAML package's `Command` aspect, and while I love how the source generator in this works when using the `ICommand` attribute, I do have one suggestion regarding the `CanExecute` functionality. To further reduce some of the boilerplate, perhaps the source generator for `ICommand` could try to detect some patterns for a `CanExecute` method or property for a command and apply that to the generated code automatically. Like for example, if I do this:

```csharp
public bool CanExecuteInvoke => this.IfTheStarsAlign;

[ICommand]
void Invoke() => this.UnalignStars();
```

It should be the same output as if I had added the `CanExecute = nameof(CanExecuteInvoke)` to the `ICommand` attribute. I don't know how feasible it would be to check for maybe `Can<command>` or `CanExecute<command>` as patterns, but those are two that I would look for, the latter being similar to what PostSharp looks for in its aspect.

**Comment by Sergio0694** [comment]
So this is an interesting idea, but I'm not entirely sure how to fit it with the rest of the features we support. One aspect of the `CanExecute` parameter in `[ICommand]` is that it supports several different signatures for target members. For instance, it can not only bind to a simple method with no parameters returning a `bool`, but also to one with a parameter, to a method with no parameters even if the command takes one (in which case the parameter is automatically ignored), to a property of type `bool` (in which case the generated code automatically wraps it with a lambda), etc.

What we could do here is to use a similar approach to that of observable properties (suggested by @sonnemaf), and have the generator do the following:

```csharp
partial bool On<COMMAND_NAME>();
```

You could then implement this in your code, and the generator would pick it up. The issue is that we can't really rely on the compiler removing this call if there are no definitions, because we'd need to pass this as a lambda to the command constructor, so we can't leverage the same trick in this case. We could potentially only conditionally generate those members and otherwise reuse the same setup, but eh, not a fan of doing that. We could also do some auto-detection like proposed, but I don't like that kind of "magic" and we're also already explicitly not doing things like that in other generators, so that would feel inconsistent. And in all cases, this approach would (1) not cover the same options as the current one, and (2) if kept as an additional option, having both would feel confusing.

So, currently leaning towards not supporting this 🤔

**Comment by skint007** [comment]
I'm using build 94 and am having issues with the ICommand canExecute param.
When I update the below Property to true, the button doesn't become clickable.
```
[ObservableProperty]
private bool _canSaveUser;

[ICommand(CanExecute = nameof(CanSaveUser))]
private async Task SaveUser()
```

**Comment by Sergio0694** [comment]
Hi @skint007, it seems you just forgot to actually notify your command, no? 🤔

Try this:

```csharp
[ObservableProperty]
[AlsoNotifyCanExecuteFor(nameof(SaveUserCommand))]
private bool _canSaveUser;

[ICommand(CanExecute = nameof(CanSaveUser))]
private async Task SaveUser()
```

This should solve your issue. We'll need to document all of these scenarios properly in our docs.

**Comment by powerdude** [comment]
@Sergio0694 this may beg the question, why can't the `ICommand` realize that the `CanExecute` value provided to it is an `ObservableProperty` and add the necessary code like `AlsoNotifyCanExecuteFor` does.  It would possibly prevent this disconnect and need for the extra wiring.

I understand one reason my be due to the timing of the generation of the property and the command, but would there be a way to delay/cache the generation until all properties/commands have been looked at?  Or have `ObservableProperty` scan and see if it is being used a value in an `ICommand`?

**Comment by skint007** [comment]
> Hi @skint007, it seems you just forgot to actually notify your command, no? 🤔
> 
> Try this:
> 
> ```cs
> [ObservableProperty]
> [AlsoNotifyCanExecuteFor(nameof(SaveUserCommand))]
> private bool _canSaveUser;
> 
> [ICommand(CanExecute = nameof(CanSaveUser))]
> private async Task SaveUser()
> ```
> 
> This should solve your issue. We'll need to document all of these scenarios properly in our docs.

Thank you that works :)

On another note, I'm not sure if this is an issue with the package or source generators in general, but there are many times where the generated code doesn't seem to show in Intellisense until I close out of VS and do a rebuild. For example, I couldn't compile because SaveUserCommand wasn't seen, until I restarted.

**Comment by Sergio0694** [comment]
@powerdude that's on purpose to make it clear what generated code is doing for properties and avoid too much "magic" going on. Additionally, no, there isn't a way to delate generators or express any kind of dependency between them, so doing that would require additional logic to "predict" what other generators would do. It's doable and we're doing that already in a couple cases (mostly to help with diagnostics), but would also be a bit of extra work.

@skint007 that's most likely an IntelliSense issue, the whole support for source generators still has some issues that will take a bit of time to solve completely. What you can do is use VS2022 Preview to get the latest version of Roslyn, and track the various issues in their repository. For instance, here's a few issues about source generators as a whole:

- https://github.com/dotnet/roslyn/issues/51810
- https://github.com/dotnet/roslyn/issues/58435
- https://github.com/dotnet/roslyn/issues/54175
- https://github.com/dotnet/roslyn/issues/52749
- https://github.com/dotnet/roslyn/issues/50451

**Comment by azthrun** [comment]
I have a very simple view model class which contains 1 observable property and 1 method.
![image](https://user-images.githubusercontent.com/26609376/160313397-8e3e6f09-a9a7-4881-9046-2aaa419113f7.png)
For some reason, the [ObservableProperty] generates the codes pretty well but not the [ICommand].

This is a .NET Maui project (preview 14) with MVVM toolkit in version 8.0.0-preview2.
Am I doing something wrong?

**Comment by Sergio0694** [comment]
Hi @azthrun, that code seems fine at a glance, so I'm not entirely sure.
Note that this issue is just for tracking progress on the whole source generators feature. Please open a new issue and follow the included template to include all necessary info (eg. VS version, etc.) and a minimal repro we can use to reproduce this on our end. It'd help a lot to investigate this and figure out if there's an issue with the package, and if so, how to fix it. Thanks! 🙂

**Comment by Sergio0694** [comment]
Hey everyone! 8.0.0 Preview 3 is out! 🚀
You can find the announcement blog post [here](https://devblogs.microsoft.com/ifdef-windows/announcing-net-community-toolkit-v8-0-0-preview-3/). We're almost there!
Thank you for all the feedbacks and comments in this issue over the past few months, they've been **incredibly** useful! 🙌

Just to pick your brain once again, I've been having some second thoughts on `[ICommand]`.
Specifically... Shouldn't it just be called `[Command]`? Thoughts? 🤔

**Comment by powerdude** [comment]
I think that is a reasonable change.  Although I have a bit of cognitive dissonance with `Observable...` and also `Relay...`.  I don't have better prefixes to use at the moment, but if you're going to rename some things, I think it's worth revisiting the whole kit and caboodle.

**Comment by wokhan** [comment]
Thanks again for your (and contributors) work! 

Regarding this:
> Just to pick your brain once again, I've been having some second thoughts on `[ICommand]`.
> Specifically... Shouldn't it just be called `[Command]`? Thoughts? 🤔
> 
> **EDIT:** folks on Discord have also suggested `[RelayCommand]`, to be consistent with the implementation type. This also makes sense given that this attribute specifically generates some kind of relay command. Additionally, it seems everyone at the very least prefers `[Command]` over `[ICommand]`. There's also a concern for `[Command]` though about it potentialy conflicting with other attributes with the same name in the future, though I'm not sure this is an actual issue.

Since the method itself is not directly used as a Command (nor Relay) but is wrapped with the actual Relay Command, I would have called it CommandHandler or something like that...
But I'll be OK with whatever you pick, I guess functionality wins over naming here 😉

**Comment by skint007** [comment]
Would it be possible to have Attributes carried over to the generated public properties?

Example:

```
[ObservableProperty]
[NotMapped]
private Person? _selectedPerson;
```
Currently it doesn't look like they are applied.

**Comment by michael-hawker** [comment]
I agree `[Command]` is going to potentially conflict with other things. `[RelayCommand]` as an attribute wouldn't conflict with `RelayCommand` the implementation though, eh? As `[RelayCommand]` is really just short for `RelayCommandAttribute`, right?

So that seems like a safe possibility.

Agree with @skint007 that it'd be nice for other attributes to carry-over, though not sure if that's possible as if an attribute is property specific for its target then it wouldn't be allowed on a field in the first place, right? Also, we wouldn't know if it's some other source generator attribute (outside of our own)? This does seem like an important gap to think about though still somehow... 🤔 But is this a general source-generator level issue?

**Comment by timunie** [comment]
I think other Attributes can be added like AlsoNotifyChangFor.

Would be 

```c#
[AttributeDecorator (nameof(MyAttribute))]
private string myField;
```

Happy coding
Tim

**Comment by Sergio0694** [comment]
> "[RelayCommand] as an attribute wouldn't conflict with RelayCommand the implementation though, eh? As [RelayCommand] is really just short for RelayCommandAttribute, right?"

Correct, that'd work just fine. It's the same reason why `[ICommand]` works today as well even though `ICommand` is also an interface. 🙂

I also share the worry about porting attributes over, and the fact they would also cause errors if the attributes only had fields as targets. To be clear, we _are_ already doing this somewhat. To be precise, if you use validation attributes on fields, we're porting them over to the generated properties too, as that was required to make it all work with validation. I'd consider that a very specific scenario though.

I'll be honest here: I'm not happy about using annotated fields, and admittedly this is just a workaround until C# gets partial properties, which would make all of this much easier to do and also for us to generate code for. If you're interested in this, please upvote and/or comment on the actual language proposal: https://github.com/dotnet/csharplang/discussions/3412. Ideally, that'd get enough traction to make it to C# at some point, and then all of this issues would just go away 🙂

**Comment by michael-hawker** [comment]
Let's talk about the _forwarding of attributes_ in the new issue here: https://github.com/CommunityToolkit/dotnet/issues/208

**Comment by CesGan** [comment]
Hello all,

I´ve got stuck on a MVVMTK0009 error similar to 
https://github.com/CommunityToolkit/dotnet/issues/8#issuecomment-1025259037

Got the error by doing the following:
```csharp
public abstract partial class ViewModelBase : ObservableObject
    {
        [ObservableProperty]
        private bool canSave;

        public abstract void Save();
    }

public partial class ViewModel : ViewModelBase
    {
        [ICommand(CanExecute = nameof(CanSave))]
        public override void Save()
        {
        }
    }
```

With this code i got the following error:

Severity	Code	Description	Project	File	Line	Suppression State
Error	MVVMTK0009	The CanExecute name must refer to a valid member, but "CanSave" has no matches in type ViewModel

**Comment by kmgallahan** [comment]
@CesGan canSave vs CanSave

**Comment by Sergio0694** [comment]
This was fixed by https://github.com/CommunityToolkit/dotnet/pull/106.
Can you try with a preview package to double check it works for you? 🙂
You can grab whatever the latest version is from: https://pkgs.dev.azure.com/dotnet/CommunityToolkit/_packaging/CommunityToolkit-MainLatest/nuget/v3/index.json.
Thank! 😄

**Comment by CesGan** [comment]
> This was fixed by #106. Can you try with a preview package to double check it works for you? 🙂 You can grab whatever the latest version is from: https://pkgs.dev.azure.com/dotnet/CommunityToolkit/_packaging/CommunityToolkit-MainLatest/nuget/v3/index.json. Thank! 😄

Hey Sergio,
The problem persists with build 157, the latest.

If i move the CanSave property to ViewModel, it works. Just doesnt work the way is set above.

**Comment by Sergio0694** [comment]
Ah, gotcha. Yeah that makes sense now that I look at that PR again, I only went through the field symbols for the containing type looking for that attribute, but not through the ones from base types as well. I'll need to loop through those as well. Would you mind opening an issue with this info (the minimal repro, etc.) so I can link that to a PR to help with tracking? Thanks! 🙂

**Comment by CesGan** [comment]
There you go @Sergio0694
#257
Hope I've done it well since I'm still not very used to this stuff.😄

**Comment by CesGan** [comment]
Hello once again,

The same problem appears now if the ViewModelBase comes from a project dependency.
Created this solution to show the problem
[https://github.com/CesGan/CesGan.CommunityToolkit.Mvvm.Test.ICommandCanExecuteAtribute](https://github.com/CesGan/CesGan.CommunityToolkit.Mvvm.Test.ICommandCanExecuteAtribute)

Using CommunityToolkit.Mvvm 8.0.0-build.165

**Comment by Sergio0694** [comment]
Hey @CesGan, thank you for the repro!
Could you please open an issue so we can track this and link it to a PR? 🙂

**Comment by Sergio0694** [comment]
Sharing notes here from the API review meeting @michael-hawker, @Arlodotexe and I just had.

We discussed the various attributes for `[ObservableProperty]` and the changes in #267.
There were several issues raised:
- We want to make the attributes consistent with how `SetProperty` methods work (eg. when using `validate: true`).
- We want to give devs the ability to just specify opt-in attributes for whole classes too.
- The names of the various attributes are a bit opinionated/not entirely clear/don't work over classes.

We ended renaming the various attributes to make them more consistent. Specifically, we dropped the "Also" prefix, and just used the exact names from the various underlying APIs/interfaces to clearly express what they're doing. That is, we have:

```csharp
[ObservableProperty]
[NotifyPropertyChangedFor("Foo")]
[NotifyCanExecuteChangedFor("Bar")]
[NotifyDataErrorInfo]
[NotifyRecipients]
private string? name;
```

Going over the changes:

- We decided to use the "Notify" prefix for all of them. This achieves two things:
  - We still have a shared prefix for all `[ObservableProperty]` modifiers (ie. these attributes don't work on their own).
  - The prefix is more consistent with the base interfaces and familiar to users (eg. [`INotifyPropertyChanged`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged), etc.).
- `AlsoNotifyChangeFor` ---> `NotifyPropertyChangedFor`. This refers to [`INotifyPropertyChanged`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifypropertychanged), plus the "For".
- `AlsoNotifyCanExecuteFor` ---> `NotifyCanExecuteChangedFor`. This refers to [`NotifyCanExecuteChanged`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.mvvm.input.irelaycommand.notifycanexecutechanged). plus the "For".
  - The "For" suffix in these cases indicates that this attribute specifically targets some other member and not this one.
- `AlsoValidateProperty` ---> `NotifyDataErrorInfo`. This refers to [`INotifyDataErrorInfo`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.inotifydataerrorinfo).
- `AlsoBroadcastChange` ---> `NotifyRecipients`. This refers to [`IRecipient<T>`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.toolkit.mvvm.Messaging.irecipient-1). This might use some more work maybe.

We also want to allow `[NotifyDataErrorInfo]` and `[NotifyRecipients]` to be used over a class. Doing so would opt-in all generated properties within that class into their respective behavior. This would still keep everything consistent with the extra functionality always being opt-in, while still giving developers wanting that as the default to do that with a single line of code.

We still have to go through over some other APIs in the next meetings, but would love some feedbacks on this 😄

**Comment by powerdude** [comment]
I like the rename of the attributes, but I think inheriting from `ObservableValidator` is opting into `INotifyDataErrorInfo` just like inheriting from `ObservableObject` is opting into `INotifyPropertyChanged`.  There shouldn't be a need to add `NotifyDataErrorInfo` on top of inheriting from `ObservableValidator`.  It doesn't make sense to have double opt-in for this use case and should allow for an opt-out.  Something like `NotifyDataErrorInfoOptOut` or `DoNotNotifyDataErrorInfo`

**Comment by michael-hawker** [comment]
> I like the rename of the attributes, but I think inheriting from ObservableValidator is opting into INotifyDataErrorInfo just like inheriting from ObservableObject is opting into INotifyPropertyChanged. There shouldn't be a need to add NotifyDataErrorInfo on top of inheriting from ObservableValidator. I think there is little value to have double opt-in for this use case for the sake of what is called consistency and should allow for an opt-out. Something like NotifyDataErrorInfoOptOut or DoNotNotifyDataErrorInfo or NotifyDataErrorInfo(false)

@powerdude we discussed this all at length as well.

There were a number of factors we went through when discussing these options:
- Looking at how developers had to write code from the existing base classes as shipped with 7.x (more detail below)
- Being consistent with expectations on how things currently work and not making things too 'magical' ✨
- Providing developers the tools they need to be in exact control of what gets generated.
- Not adding extra behaviors with overhead appearing by default (be explicit about the cost of adding extra code)
- Making sure the patterns were applied uniformly across all the implementations of the base classes so that one class didn't have a default behavior different than another.
- Ensuring that we make sure to have good samples and docs here, as these are more advanced scenarios.

We thought allowing the attributes at the class level was a good compromise as well to align to all these principles as well.

## ObservableObject

To break it down from how we looked at it; today with an `ObservableObject`, a developer still needs to do the following for `INotifyPropertyChanged` to work as expected (over a standard property):

```cs
public class MyClass : ObservableObject
{
    private string foo;
    public string Foo
    {
       get => foo;
       set => SetProperty(ref foo, value);
    }
```

With source generators this is simplified to:

```cs
public partial class MyClass : ObservableObject
{
    [ObservableProperty]
    private string foo;
```

Nothing prevents you from adding a regular property to an `ObservableObject` class:

```cs
public class MyClass : ObservableObject
{
    public string Foo { get; set; }
```

But this won't notify anyone of changes, the developer still has to modify code to make things work even when inheriting from the `ObservableObject` base class. This was the crux of the rest of our process which cascaded to our proposed solution and expanded into the other base classes below. I'll complete these examples to show how it will make everything consistent.

## ObservableRecipient

If we look at `ObservableRecipient` the `broadcast` option is opt-in from the current implementation's base class:

```cs
public class MyClass : ObservableRecipient
{
    private string foo;
    public string Foo
    {
       get => foo;
       set => SetProperty(ref foo, value, broadcast: true);
    }
```

This will map with source generators to:

```cs
public partial class MyClass : ObservableRecipient
{
    [ObservableProperty]
    [NotifyRecipients] // This represents the extra parameter now, that was still required in the non-sg implementation
    private string foo;
```

## ObservableValidator

If we look at `ObservableValidator` the `validate` option is opt-in from the current implementation's base class:

```cs
public class MyClass : ObservableValidator
{
    private string foo;

    [Required]
    public string Foo
    {
       get => foo;
       set => SetProperty(ref foo, value, validate: true);
    }
```

This will map with source generators to:

```cs
public partial class MyClass : ObservableValidator
{
    [ObservableProperty]   
    [NotifyDataErrorInfo] // This represents the extra parameter now, that was still required in the non-sg implementation
    [Required]
    private string foo;
```

**Comment by powerdude** [comment]
hi @michael-hawker , 

thanks for the details.  I think it highlights for me one disagreement I have with the design that I'll share here.  This code:

```csharp
public class MyClass : ObservableRecipient
{
    private string foo;
    public string Foo
    {
       get => foo;
       set => SetProperty(ref foo, value, broadcast: true);
    }
```

should probably be just:

```csharp
public class MyClass : ObservableRecipient
{
    private string foo;
    public string Foo
    {
       get => foo;
       set => SetProperty(ref foo, value);
    }
```

The fact that you're in a new class should mean that `SetProperty`, if it was virtual, should just be overridden to add the new behavior as it should come for free because you've inherited from `ObservableRecipient`.  I would argue the default behavior should be to broadcast the property if using `ObservableProperty`.  In fact, it was my initial thought when I first started using the library and I would argue others would think the same.  I don't think one would think it to be magic because they intentionally inherited from the `ObservableRecipient`.  Again, this feels like double opt-in which I think will more than likely frustrate new developers and increase the friction to entry with people reporting bugs because they **_forgot_** to add the extra attribute..  I think the same can be said for `ObservableValidator`.  The combination of inheriting from a specific object and using `ObservableProperty` should unlock the features of the base class used, otherwise, why inherit from the base class.  

I hope some more consideration can be given to this, but I understand if it can't.  Code generators are **_magic_** and the library shouldn't be afraid of embracing that and doing more for the developer when it can.  The exceptions are what the additional attributes should be for, i.e `NotifyCanExecuteFor`, `NotifyPropertyChangedFor`, etc.

**Comment by michael-hawker** [comment]
Thanks @powerdude, that's a valid argument! We did discuss it to! 😉 (I didn't bring it up in my big comment above as it was a bit orthogonal to the rest of our discussion/reasoning.)

There was a technical thing we did talk about. The main thing is the original design of `ObservableObject` didn't make the `SetProperty` method `virtual`, so both `ObservableRecipient` and `ObservableValidator` use an overload to add their new behaviors compared to overriding the default behavior.

We could change this as it is a major release, but that'd involve a lot of work and also 'silently' change the default behavior for folks updating. We felt this would be pretty bad if everything suddenly flipped on someone implementing these patterns already. Plus, it kind of fell back into the principle of not adding extra code to a method unless requested, as different folks have different scenarios where they don't want to broadcast every change or only validate things when the user clicks a button.

We don't really want to detect what we think the expected behavior should be, so this seemed like the best approach originally when we created these APIs, and I think that's helping lead us in the same direction now with the source generators.

**Comment by Sergio0694** [comment]
Almost there with the (hopefully last) next preview! 🎉

The last changes that need to go in are the tweaks to async commands, see here: https://github.com/CommunityToolkit/dotnet/pull/292.

Specifically, we're adding these new APIs:

```csharp
namespace CommunityToolkit.Mvvm.Input;

[Flags]
public enum AsyncRelayCommandOptions
{
    None,
    AllowConcurrentExecutions,
    FlowExceptionsToTaskScheduler
}

public sealed class RelayCommandAttribute
{
    public bool FlowExceptionsToTaskScheduler { get; init; }
}
```

And changing the default behavior of async commands to await the resulting task, which also causes them to re-throw the exception in case the wrapped delegate fails. This also makes them more consistent with how normal commands work. In order to get back to the previous behavior and get the ability to manually inspect even faulted tasks, you can use the new `FlowExceptionsToTaskScheduler` option, which makes the commands not await the resulting task.

**Comment by Sergio0694** [comment]
**PSA:** 8.0.0 Preview 4 is now live on NuGet! 🎉

See release notes [here](https://github.com/CommunityToolkit/dotnet/releases/tag/v8.0.0-preview4).
This version includes all the recent changes that have been discussed. Let me know how it goes if you try it out!
If nothing else pops up, this looks like a good release candidate 😄

**Comment by virzak** [comment]
Thanks @Sergio0694,

[ICommand] => [RelayCommand] isn't immediately visible in the release notes, but is important for those updating from preview 3.

Also logged what seems to be a bug #294

**Comment by Sergio0694** [comment]
> "there might be cases, where a user would like to specify it"

That doesn't seem like a very common scenario to me, in fact it has never been brought up before either.
Also I personally dislike having identifiers be specified as a string for generated members, and we've tried to avoid that.
The current pattern ensures there's a predictable and consistent behavior with respect to command/property names.
Overall, I don't think there's enough to make such an addition be worth the extra complexity 🙂

**Comment by pekspro** [comment]
I just wrote this code:

``` C#
[ObservableProperty]
private ulong _BytesDownloaded;

partial void OnBytesDownloadedChanged(ulong value)
{
    DoStuff();
}
```

Works fine. But it would be nice if we could do something like this instead:

``` C#
[ObservableProperty]
[CallWhenChanged(nameof(DoStuff))]
private ulong _BytesDownloaded;
```

Or it makes it unnecessarily complicated 😁

**Comment by timunie** [comment]
> I just wrote this code:
> 
> ```cs
> [ObservableProperty]
> private ulong _BytesDownloaded;
> 
> partial void OnBytesDownloadedChanged(ulong value)
> {
>     DoStuff();
> }
> ```
> 
> Works fine. But it would be nice if we could do something like this instead:
> 
> ```cs
> [ObservableProperty]
> [CallWhenChanged(nameof(DoStuff))]
> private ulong _BytesDownloaded;
> ```
> 
> Or it makes it unnecessarily complicated 😁

From my point of view this can get quite complex. One may want to call an `async void / Task`  or one may want / need to pass paramters. One may need to handle the result of `DoStuff()`. ...

**Comment by chrisk414** [comment]
Has anyone mentioned about DependencyProperty source generator? It is possible to add it to the roadmap? It is one of the properties that require ugly boilerplates.

**Comment by chrisk414** [comment]
> The MVVM Toolkit is UX Framework agnostic, so it wouldn't fit in the existing `CommunityToolkit.MVVM` .NET Standard package.

It looks like the new-kinds-in-the-block (MAUI) decided to call it BindableProperty, breaking the nomenclatures. Too bad and perhaps the code generator is smart enough to be able to generate different codes for a different framework to make it framework agnostic. 
But it's cool to know HavenDV already has the solution. 

One thing I was curious about is, how will the debugging work with so many codes are generated behind? Will breakpoint work for instance? If I want to break on the Setter, I'm not even sure where to put the breakpoint on. ^^
Perhaps that's the price we have to pay.

**Comment by michael-hawker** [comment]
@chrisk414 underneath `Dependencies` in your project in VS, you can open up `Analyzers` find the source generator that's generating code, and open the generated file, I believe as long as 'Just My Code' is disabled then it should allow you to set breakpoints and stop in the generated code there, like anything else.

**Comment by HavenDV** [comment]
> It looks like the new-kinds-in-the-block (MAUI) decided to call it BindableProperty, breaking the nomenclatures. Too bad and perhaps the code generator is smart enough to be able to generate different codes for a different framework to make it framework agnostic. But it's cool to know HavenDV already has the solution.
> 
> One thing I was curious about is, how will the debugging work with so many codes are generated behind? Will breakpoint work for instance? If I want to break on the Setter, I'm not even sure where to put the breakpoint on. ^^ Perhaps that's the price we have to pay.

Yes, MAUI uses its own classes to create Dependency properties and the generator doesn't support that at the moment. I'll add partial MAUI support today, but I'll need help testing it out because I don't have MAUI projects right now. Also at the moment I don't know how to ensure the correct configuration for MAUI and check diagnostics using Roslyn, so the work will be done blindly. On the plus side, MAUI has one method signature for all kinds of properties, which makes the task easier.

**Comment by CBenghi** [comment]
Hello, 
I've got a little suggestion for even smarter code generation; consider the following code, where a command can execute depends on the specified member `CanDoSomething`:

``` cs
public partial class Feedback : ObservableObject
{
	[ObservableProperty]
	[NotifyPropertyChangedFor(nameof(CanDoSomething))]
	[NotifyCanExecuteChangedFor(nameof(DoSomethingCommand))] // redundant, in the definition of RelayCommand we know that CanDoSomething should trigger CanExecuteChanged on the command
	private string fileName1 = "";

	[ObservableProperty]
	[NotifyPropertyChangedFor(nameof(CanDoSomething))]
	[NotifyCanExecuteChangedFor(nameof(DoSomethingCommand))] // redundant, in the definition of RelayCommand we know that CanDoSomething should trigger CanExecuteChanged on the command
	private string fileName2 = "";

	public bool CanDoSomething => File.Exists(fileName1) && File.Exists(fileName2);
		
	[RelayCommand(CanExecute = nameof(CanDoSomething))]
	public void DoSomething() {}
}
```

My suggestion is to automate the `NotifyCanExecuteChangedFor`, which can be inferred because we know the command depends on changes on `CanDoSomething`, so every automation that changes `CanDoSomething` should trigger it.

I know it's only two lines saved in this scenario, but on more complex viewmodels it would allow the developers to relax and know that whenever `CanDoSomething` is changed the command's `CanExecute` is re-evaluated.

If you are interested in the idea I'd be happy to propose a PR.

**Comment by kamil-cupial** [comment]
Hi,

It looks like source generators produce errors on newer SKDs. When building with .NET6 SDKs newer than 6.0.101. I get a bunch (500+) of compilation errors. This does not occur if I build with 6.0.101.
Tested with SDK 6.0.3, 6.0.107 and preview-4 and 3.

All errors look like these:

```
[project path]\CommunityToolkit.Mvvm.SourceGenerators\CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator\[namespace and file name].cs(170,22): error CS0111:
Type '[class name]' already defines a member called 'On[property]Changed' with the same parameter types [[project path]\[project name]_wwuqb4gk_wpftmp.csproj]
```

`error CS0756: A partial method may not have multiple defining declarations`

` error CS0535: '[class].TaskNotifier<T>' does not implement interface member '[class].ITaskNotifier<Task<T>>.Task'`

`error CS0102: The type '[class]' already contains a definition for 'ITaskNotifier'`

This is not just about ITaskNotifier but I get duplicates about OnChanged/Changing methods etc. too.

**Comment by JorgeCandeias** [comment]
> It looks like source generators produce errors on newer SKDs.

This is an sdk regression tracked here (https://github.com/dotnet/wpf/issues/6792), identified here (https://github.com/dotnet/wpf/issues/6792#issuecomment-1183471065) and there are workarounds here (https://github.com/dotnet/wpf/issues/6792#issuecomment-1183471011) and here (https://github.com/dotnet/wpf/issues/6792#issuecomment-1183633402) until the fix (https://github.com/dotnet/wpf/pull/6793) is available.

**Comment by MisinformedDNA** [comment]
From what I'm reading, these attributes are supposed to be for 8.x, but I am seeing some of these in 7.1. Are these supposed to be there? Are they stable to use?

![image](https://user-images.githubusercontent.com/1784452/180090301-e3fcab10-47b8-48c0-9d37-7eccea6571fc.png)
![image](https://user-images.githubusercontent.com/1784452/180090390-6efdd277-f086-4ed8-9005-a0ae63e5feed.png)

**Comment by michael-hawker** [comment]
@MisinformedDNA we shipped a preview of the source generators in 7.1.x. However, they're built on top of the previous version of the Source Generator API from .NET, so they fall over with larger loads. In retrospect we probably should have separated them out as a separate package, but we didn't quite realize the impact at the time.

These new generators with 8.0 have been refined and completely rebuilt with the new Incremental generator API, so they'll work under any sized project. I know Sergio is planning a release in the near future, and I don't think there's any planned breaking changes since the last preview release.

Did I miss anything @Sergio0694?

**Comment by Sergio0694** [comment]
Nope, that's exactly right. I would **strongly recommend not to use** the attributes in 7.1. They're not supported for production. If you want to use the generators, do use `CommunityToolkit.Mvvm 8.0.0`, which in Preview 4 is basically RTM anyway. We're hopefully release a stable version of it by the end of the month 🙂

**Comment by MisinformedDNA** [comment]
The preview version doesn't exist under the `Microsoft.Toolkit.Mvvm` NuGet package. After reading the [roadmap](https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/4488), it seems like everything is moving to the `CommunityToolkit` prefix, is that correct?

**Comment by Sergio0694** [comment]
Yup that's correct, the new package/namespace will be `CommunityToolkit` 🙂
I'll document this in the blog post too, and will deprecate the `Microsoft.Toolkit.*` packages too to avoid confusion.

**NOTE:** the package is still published and maintained by Microsoft, it's just a name change.

**Comment by MisinformedDNA** [comment]
Are all the `CommunityToolkit` repos maintained by Microsoft? Or just the MVVM one? Seems weird to get rid of the `Microsoft.Toolkit` prefix. I read that it is due to reserving NuGet package prefixes, but if Microsoft owns it, I don't see the problem.

**Comment by Sergio0694** [comment]
> "Are all the CommunityToolkit repos maintained by Microsoft?"

As far as I know, yes. For sure the .NET, Windows and MAUI Community Toolkit-s are.

> "I read that it is due to reserving NuGet package prefixes, but if Microsoft owns it, I don't see the problem."

It just makes things simpler for us for several reasons we don't have to go into 🙂
The publisher (on NuGet as well) is still Microsoft anyway. It's really just a name change, that's it.
Like, we're also migrating to the new CommunityToolkit.* repos in our first party apps (eg. the Store). Nothing changed 😄

**Comment by timunie** [comment]
> It just makes things simpler for us for several reasons we don't have to go into 🙂

Just to add from my side: I think for people like me who are using Avalonia or any other not MS UI framework may wonder if they can use something that has Microsoft.* in the name. So I like the new name tbh. 

Happy coding
Tim

**Comment by chylex** [comment]
Could it be possible for `[ObservableProperty]` to generate a non-public setter? I have several models with properties that have a `private` or `internal` setter, and I don't see a way to replace the `public` setter that `[ObservableProperty]` generates now.

**Comment by timunie** [comment]
> Could it be possible for `[ObservableProperty]` to generate a non-public setter? I have several models with properties that have a `private` or `internal` setter, and I don't see a way to replace the `public` setter that `[ObservableProperty]` generates now.

That would be useful for read-only properties, yes. I vote for it :-).  Let me add: At least in Avalonia one can bind to internal properties. So maybe properties can also be made internal on request. I don't use this feature of Avalonia until today, but probably it's worth to mention here.

**Comment by chylex** [comment]
> > Could it be possible for `[ObservableProperty]` to generate a non-public setter? I have several models with properties that have a `private` or `internal` setter, and I don't see a way to replace the `public` setter that `[ObservableProperty]` generates now.
> 
> That would be useful for read-only properties, yes. I vote for it :-). Let me add: At least in Avalonia one can bind to internal properties. So maybe properties can also be made internal on request. I don't use this feature of Avalonia until today, but probably it's worth to mention here.

Both the getter and setter visibility could be properties of the attribute, maybe something like:
```csharp
[ObservableProperty(Getter = Visibility.Internal, Setter = Visibility.Private)]
```

If there's interest and we can work out the details of usage and naming, I can make a PR - or just leave it to the maintainers, whichever way would be easier for them.

**Comment by pekspro** [comment]
@chylex, this has been suggested earlier:

https://github.com/CommunityToolkit/dotnet/issues/291

(spoiler alert: maybe in C# 12)

**Comment by chylex** [comment]
> @chylex, this has been suggested earlier:
> 
> #291
> 
> (spoiler alert: maybe in C# 12)

Quoting the reply from #291:

> This is not supported and not planned. We want to support this properly in C# 12 via partial properties. For now the plan is to work on a proposal for that to add to the working set, and I don't think it'd be a good idea to hack a clunkier alternative for this just for the short term until C# 12 was out (which would be november 2023). That would also solve a whole lot of other problems related to the fact we're currently using fields in the first place 🙂

Personally I disagree that november 2023 is 'short term', and if the solution are partial properties instead of fields then that sounds like it's going to be a breaking change for users of Community Toolkit anyway.

I might try to implement this anyway. If the maintainers don't want to accept a PR and officially support this feature, which would be understandable, I can just put it on a self-hosted NuGet repository and use it in my own projects.

**Comment by Sergio0694** [comment]
> "if the solution are partial properties instead of fields then that sounds like it's going to be a breaking change for users of Community Toolkit anyway"

I didn't say support for fields would be removed. We can very well consider just leaving it there side by side by properties, or at least keeping it there for one release cycle to give everyone time to migrate to properties. Plus, this is not a show stopper. If you absolutely need a private setter for the time being, you can use manual properties in those cases, and still use generated properties everywhere else 🙂

> "I might try to implement this anyway"

Don't do this. As the PR template says, PRs without an approved linked proposal will just be closed.

**Comment by michael-hawker** [comment]
> The preview version doesn't exist under the `Microsoft.Toolkit.Mvvm` NuGet package. After reading the [roadmap](https://github.com/CommunityToolkit/WindowsCommunityToolkit/issues/4488), it seems like everything is moving to the `CommunityToolkit` prefix, is that correct?

You can find more information about this from [one of our blog posts](https://devblogs.microsoft.com/ifdef-windows/windows-community-toolkit-for-project-reunion-0-5/#package-identity-and-renaming), to quote:

> ... having our [own] package identity allows us to better maintain our packages and their ownership with the community. In the future, it also allows us more flexibility to work with other groups and technologies to share branding and ownership of packages.

**Comment by todor-dk** [comment]
> Don't do this. As the PR template says, PRs without an approved linked proposal will just be closed.

Fair enough. And it is now too late to to get this wish in the release, what would have been the best procedure to get the change in an official release. The change that @chylex did is very desirable and I've considered it myself.

In other words, how would one have created an **approved linked proposal**, so a PR can get through.

**Comment by chylex** [comment]
A proposal already existed and was denied, so I don't think there was any way for a PR to get through. If the maintainers decided to allow the proposal, I would still be interested in making a PR for the benefit of others, but for now I would recommend what I did - build a custom version from my fork, and create a private NuGet repository.

If it helps, I can upload my NuGet repository somewhere, since it is made entirely of static files so you'd just need to upload it to a web server. Or use my repository, but I can't guarantee its uptime or which versions of the toolkit would be available. Details would be best discussed in [my fork's discussions](https://github.com/chylex/.NET-Community-Toolkit/discussions).

**Comment by Shujee** [comment]
Would it be possible for `CanExecutre` parameter of `RelayCommand` attribute to directly take a boolean lambda in addition to the name of a bool function? More often than not, `CanExecute` is one-liner expression and writing a dedicated method for it seems verbose. and more work to manage.

**Comment by MisinformedDNA** [comment]
This issue is closed. Create a new issue.
